src/thread.rs

```diff
@@ -0,0 +1,583 @@
+                .iter()
+                .all(|thread| thread.state != ThreadState::BlockedOnJoin(joined_thread_id)),
+            "a joinable thread has threads waiting for its termination"
+        );
+        // Mark the joined thread as being joined so that we detect if other
+        // threads try to join it.
+        self.threads[joined_thread_id].join_status = ThreadJoinStatus::Joined;
+        if self.threads[joined_thread_id].state != ThreadState::Terminated {
+            // The joined thread is still running, we need to wait for it.
+            self.active_thread_mut().state = ThreadState::BlockedOnJoin(joined_thread_id);
+            trace!(
+                "{:?} blocked on {:?} when trying to join",
+                self.active_thread,
+                joined_thread_id
+            );
+        }
+        Ok(())
+    }
+
+    /// Set the name of the active thread.
+    fn set_thread_name(&mut self, new_thread_name: Vec<u8>) {
+        self.active_thread_mut().thread_name = Some(new_thread_name);
+    }
+
+    /// Get the name of the active thread.
+    fn get_thread_name(&mut self) -> InterpResult<'tcx, Vec<u8>> {
+        if let Some(ref thread_name) = self.active_thread_mut().thread_name {
+            Ok(thread_name.clone())
+        } else {
+            throw_ub_format!("thread {:?} has no name set", self.active_thread)
+        }
+    }
+
+    /// Allocate a new blockset id.
+    fn create_blockset(&mut self) -> BlockSetId {
+        self.blockset_counter = self.blockset_counter.checked_add(1).unwrap();
+        self.blockset_counter.into()
+    }
+
+    /// Block the currently active thread and put it into the given blockset.
+    fn block_active_thread(&mut self, set: BlockSetId) {
+        let state = &mut self.active_thread_mut().state;
+        assert_eq!(*state, ThreadState::Enabled);
+        *state = ThreadState::Blocked(set);
+    }
+
+    /// Unblock any one thread from the given blockset if it contains at least
+    /// one. Return the id of the unblocked thread.
+    fn unblock_some_thread(&mut self, set: BlockSetId) -> Option<ThreadId> {
+        for (id, thread) in self.threads.iter_enumerated_mut() {
+            if thread.state == ThreadState::Blocked(set) {
+                trace!("unblocking {:?} in blockset {:?}", id, set);
+                thread.state = ThreadState::Enabled;
+                return Some(id);
+            }
+        }
+        None
+    }
+
+    /// Change the active thread to some enabled thread.
+    fn yield_active_thread(&mut self) {
+        self.yield_active_thread = true;
+    }
+
+    /// Decide which action to take next and on which thread.
+    ///
+    /// The currently implemented scheduling policy is the one that is commonly
+    /// used in stateless model checkers such as Loom: run the active thread as
+    /// long as we can and switch only when we have to (the active thread was
+    /// blocked, terminated, or was explicitly asked to be preempted).
```

Comment:
```suggestion
    /// blocked, terminated, or has explicitly asked to be preempted).
```
You can only preempt yourself, is that correct?

Reply:
Yes. At least currently.